core/jvm/src/main/scala/fs2/compress.scala

```diff
@@ -75,197 +105,658 @@ object compress {
+    val crc16 = Array[Byte](
+      (crc32Value & 0xff).toByte,
+      ((crc32Value >> 8) & 0xff).toByte
+    )
+    Stream.chunk(asChunkBytes(header)) ++
+      fileNameEncoded
+        .map(bytes => Stream.chunk(asChunkBytes(bytes)) ++ Stream.emit(zeroByte))
+        .getOrElse(Stream.empty) ++
+      commentEncoded
+        .map(bytes => Stream.chunk(asChunkBytes(bytes)) ++ Stream.emit(zeroByte))
+        .getOrElse(Stream.empty) ++
+      Stream.chunk(asChunkBytes(crc16))
+  }
 
-        body ++ trailer
+  private def _gzip_trailer[F[_]](deflater: Deflater, crc32: CRC32): Stream[F, Byte] =
+    Stream.suspend {
+      // See RFC 1952: https://www.ietf.org/rfc/rfc1952.txt
+      val (crc32Value, bytesIn) = crc32.synchronized {
+        (crc32.getValue, deflater.getTotalIn)
       }
+      val trailer = Array[Byte](
+        (crc32Value & 0xff).toByte, // CRC-32: Cyclic Redundancy Check
+        ((crc32Value >> 8) & 0xff).toByte,
+        ((crc32Value >> 16) & 0xff).toByte,
+        ((crc32Value >> 24) & 0xff).toByte,
+        (bytesIn & 0xff).toByte, // ISIZE: Input size
+        ((bytesIn >> 8) & 0xff).toByte,
+        ((bytesIn >> 16) & 0xff).toByte,
+        ((bytesIn >> 24) & 0xff).toByte
+      )
+      Stream.chunk(asChunkBytes(trailer))
+    }
 
   /**
-    * Returns a pipe that incrementally decompresses input according to the GZIP
-    * format. Any errors in decompression will be sequenced as exceptions into the
-    * output stream. The implementation of this pipe delegates directly to
-    * `GZIPInputStream`. Despite this, decompression is still handled in a streaming
-    * and async fashion without any thread blockage. Under the surface, this is
-    * handled by enqueueing chunks into a special type of byte array InputStream
-    * which throws exceptions when exhausted rather than blocking. These signal
-    * exceptions are caught by the pipe and treated as an async suspension. Thus,
-    * there are no issues with arbitrarily-framed data and chunk boundaries. Also
-    * note that there is almost no performance impact from these exceptions, due
-    * to the way that the JVM handles throw/catch.
-    *
-    * The chunk size here is actually really important. If you set it to be too
-    * small, then there will be insufficient buffer space for `GZIPInputStream` to
-    * read the GZIP header preamble. This can result in repeated, non-progressing
-    * async suspensions. This case is caught internally and will be raised as an
-    * exception (`NonProgressiveDecompressionException`) within the output stream.
-    * Under normal circumstances, you shouldn't have to worry about this. Just, uh,
-    * don't set the buffer size to something tiny. Matching the input stream largest
-    * chunk size, or roughly 8 KB (whichever is larger) is a good rule of thumb.
+    * Returns a `Pipe` that gunzips (decompresses) its input elements as defined by RFC 1952:
+    * https://www.ietf.org/rfc/rfc1952.txt
```

Comment:
Ditto above

Reply:
Original documentation restored with the addition of the RFC link and the removal of the references to GZIPInputStream.

This gunzip implementation doesn't appear to as sensitive to small buffer sizes. I've had success with chunk sizes of as little as 64 bytes, but it's difficult to determine the minimum buffer size for decompressing all possible input. With a quick online search I managed to find some C# documentation that recommended a minimum 128 byte buffer, but this is tenuous at best and I wouldn't trust much other than actual implementation documentation.

Furthermore, I found that for this implementation it didn't make much different to double the size of the input buffer as the size of the output buffer because the output is chunked out and it's difficult to predict the input compression ratio.